sys/firmware/firmware.c

```diff
@@ -0,0 +1,246 @@
+    memcpy(sm, ((unsigned char *)metadata) + FIRMWARE_SIGN_BYTES, crypto_sign_BYTES);
+    memcpy(sm + crypto_sign_BYTES, metadata, FIRMWARE_SIGN_BYTES);
+
+    unsigned char m[FIRMWARE_SIGN_BYTES + crypto_sign_BYTES];
+    unsigned long long mlen;
+    int res = crypto_sign_open(m, &mlen, sm, FIRMWARE_SIGN_BYTES + crypto_sign_BYTES, pk);
+    if (res) {
+        printf("%s: metadata signature invalid\n", __func__);
+    }
+    return res;
+}
+
+uint32_t firmware_metadata_checksum(firmware_metadata_t *metadata)
+{
+    return fletcher32((uint16_t*)metadata, FIRMWARE_CHECKSUM_LEN / 2);
+}
+
+int firmware_sign_metadata(firmware_metadata_t *metadata, unsigned char *sk)
+{
+    unsigned char sm[FIRMWARE_SIGN_BYTES + crypto_sign_BYTES];
+    unsigned long long smlen;
+    crypto_sign(sm, &smlen, (unsigned char*)metadata, FIRMWARE_SIGN_BYTES, sk);
+    memcpy(metadata->sig, sm, crypto_sign_BYTES);
+    return 0;
+}
+
+#ifdef RIOT_VERSION
+static const unsigned _firmware_slot_start[] = {
+    CPU_FLASH_BASE,
+    CPU_FLASH_BASE + FIRMWARE_SLOT0_SIZE,
+    CPU_FLASH_BASE + FIRMWARE_SLOT0_SIZE + FIRMWARE_SLOT1_SIZE
+};
+
+void firmware_jump_to_image(firmware_metadata_t *metadata)
+{
+    uint32_t addr = (unsigned)metadata + (unsigned)sizeof(firmware_metadata_t);
+    printf("[bootloader] jumping to %#lx\n", addr);
+    cpu_jump_to_image(addr);
+}
+
+int firmware_current_slot(void)
+{
+    unsigned base_addr = cpu_get_image_baseaddr() - sizeof(firmware_metadata_t);
+
+    for (unsigned i = 1; i < FIRMWARE_NUM_SLOTS; i++) {
+        if (base_addr == _firmware_slot_start[i]) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+firmware_metadata_t *firmware_get_metadata(int slot) {
+    assert(slot < FIRMWARE_NUM_SLOTS);
+    return (firmware_metadata_t *)_firmware_slot_start[slot];
+}
+
+unsigned firmware_get_image_startaddr(unsigned slot)
+{
+    assert(slot < FIRMWARE_NUM_SLOTS);
+    return _firmware_slot_start[slot] + sizeof(firmware_metadata_t);
+}
+
+uint32_t firmware_get_slot_page(uint8_t slot)
+{
+    return cpu_get_slot_page(slot);
+}
+
+void firmware_jump_to_slot(unsigned slot)
+{
+    firmware_jump_to_image(firmware_get_metadata(slot));
+}
+
+void firmware_dump_slot_addrs(void)
+{
+    for (unsigned i = 1; i < FIRMWARE_NUM_SLOTS; i++) {
+        printf("slot %u: metadata: 0x%08x image: 0x%08x\n", i, \
+                (unsigned)_firmware_slot_start[i], firmware_get_image_startaddr(i));
+    }
+}
+
+uint32_t firmware_get_slot_address(uint8_t slot)
+{
+    return cpu_get_slot_address(slot);
+}
+
+int firmware_validate_int_slot(uint8_t slot)
+{
+    firmware_metadata_t metadata;
+    int result;
+
+    assert(slot < FIRMWARE_NUM_SLOTS);
+
+    memcpy(&metadata, firmware_get_metadata(slot), sizeof(firmware_metadata_t));
+    result = firmware_validate_metadata_signature(&metadata, ed25519_pub);
+
+    if (result == -1) {
+        printf("[firmware] Signature check failed!\n");
```

Comment:
I suggest to  modify to 

```c
printf("[firmware] Signature check failed! (slot %u)\n",slot );
```
This can help to identify the wrong slot if the node already rebooted.